Content.Server/Pinpointer/NavMapSystem.cs

```diff
@@ -67,69 +78,78 @@ private void OnStationInit(StationGridAddedEvent ev)
     private void OnNavMapSplit(ref GridSplitEvent args)
     {
-        if (!TryComp(args.Grid, out NavMapComponent? comp))
+        if (!_navQuery.TryComp(args.Grid, out var comp))
             return;
 
-        var gridQuery = GetEntityQuery<MapGridComponent>();
-
         foreach (var grid in args.NewGrids)
         {
             var newComp = EnsureComp<NavMapComponent>(grid);
-            RefreshGrid(grid, newComp, gridQuery.GetComponent(grid));
+            RefreshGrid(grid, newComp, _gridQuery.GetComponent(grid));
+        }
+
+        RefreshGrid(args.Grid, comp, _gridQuery.GetComponent(args.Grid));
+    }
+
+    private NavMapChunk EnsureChunk(NavMapComponent component, Vector2i origin)
+    {
+        if (!component.Chunks.TryGetValue(origin, out var chunk))
+        {
+            chunk = new(origin);
+            component.Chunks[origin] = chunk;
         }
 
-        RefreshGrid(args.Grid, comp, gridQuery.GetComponent(args.Grid));
+        return chunk;
     }
 
     private void OnTileChanged(ref TileChangedEvent ev)
     {
-        if (!TryComp<NavMapComponent>(ev.NewTile.GridUid, out var navMap))
+        if (!ev.EmptyChanged || !_navQuery.TryComp(ev.NewTile.GridUid, out var navMap))
             return;
 
         var tile = ev.NewTile.GridIndices;
         var chunkOrigin = SharedMapSystem.GetChunkIndices(tile, ChunkSize);
 
-        if (!navMap.Chunks.TryGetValue((NavMapChunkType.Floor, chunkOrigin), out var chunk))
-            chunk = new(chunkOrigin);
+        var chunk = EnsureChunk(navMap, chunkOrigin);
+        chunk.EnsureType(NavMapChunkType.Floor);
 
         // This could be easily replaced in the future to accommodate diagonal tiles
-        if (ev.NewTile.IsSpace())
-            chunk = UnsetAllEdgesForChunkTile(chunk, tile);
-
+        if (ev.NewTile.IsSpace(_tileDefManager))
+            UnsetAllEdgesForChunkTile(chunk, tile, NavMapChunkType.Floor);
         else
-            chunk = SetAllEdgesForChunkTile(chunk, tile);
-
-        chunk.LastUpdate = _gameTiming.CurTick;
-        navMap.Chunks[(NavMapChunkType.Floor, chunkOrigin)] = chunk;
+            SetAllEdgesForChunkTile(chunk, tile, NavMapChunkType.Floor);
 
-        Dirty(ev.NewTile.GridUid, navMap);
+        if (!PruneEmpty((ev.NewTile.GridUid, navMap), chunk))
+            DirtyChunk((ev.NewTile.GridUid, navMap), chunk);
     }
 
-    private void OnAirtightChanged(ref AirtightChanged ev)
+    private void DirtyChunk(Entity<NavMapComponent> entity, NavMapChunk chunk)
     {
-        var gridUid = ev.Position.Grid;
-
-        if (!TryComp<NavMapComponent>(gridUid, out var navMap) ||
-            !TryComp<MapGridComponent>(gridUid, out var mapGrid))
+        if (chunk.LastUpdate == _gameTiming.CurTick)
             return;
 
-        // Refresh the affected tile
-        var tile = ev.Position.Tile;
-        var chunkOrigin = SharedMapSystem.GetChunkIndices(tile, ChunkSize);
+        chunk.LastUpdate = _gameTiming.CurTick;
+        Dirty(entity);
+    }
```

Comment:
Is this supposed to do something more? Dirty(EntityUid) already checks for LastModifiedTick on the component.

https://github.com/space-wizards/RobustToolbox/blob/25211e37819a2b11e7160038005259aafe8a64f0/Robust.Shared/GameObjects/EntityManager.cs#L366-L381

Reply:
Chunk isn't a component itself, this is just in case we might have a bunch of airtight changes in the same chunk on a tick (e.g. explosion) we don't try and dirty the component as well, seemed easier considering we already had it.